File: lib/core/services/tensorflow_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';
import 'package:tflite_flutter/tflite_flutter.dart';
import 'package:image/image.dart' as img;
import '../config/ai_model_config.dart';

// Top-level function for isolate-based image preprocessing
// This runs in a separate thread to avoid blocking the UI
Future<Float32List?> _preprocessImageInIsolate(Map<String, dynamic> params) async {
  try {
    final String imagePath = params['imagePath'];
    final int modelInputSize = params['modelInputSize'];
    final int modelChannels = params['modelChannels'];

    // Read and decode image
    final imageFile = File(imagePath);
    final imageBytes = await imageFile.readAsBytes();
    final image = img.decodeImage(imageBytes);

    if (image == null) {
      debugPrint('❌ Failed to decode image in isolate');
      return null;
    }

    // Resize to model input size
    final resizedImage = img.copyResize(
      image,
      width: modelInputSize,
      height: modelInputSize,
      interpolation: img.Interpolation.linear,
    );

    // Convert to Float32List and normalize
    // Using direct buffer access for better performance
    final inputSize = modelInputSize * modelInputSize * modelChannels;
    final input = Float32List(inputSize);

    // Optimized pixel processing
    int pixelIndex = 0;
    for (int y = 0; y < modelInputSize; y++) {
      for (int x = 0; x < modelInputSize; x++) {
        final pixel = resizedImage.getPixel(x, y);

        // Extract RGB values and normalize to [-1, 1] range
        input[pixelIndex++] = (pixel.r / 127.5) - 1.0;
        input[pixelIndex++] = (pixel.g / 127.5) - 1.0;
        input[pixelIndex++] = (pixel.b / 127.5) - 1.0;
      }
    }

    return input;
  } catch (e) {
    debugPrint('❌ Image preprocessing failed in isolate: $e');
    return null;
  }
}

class TensorFlowService {
  static Interpreter? _interpreter;
  static bool _isModelLoaded = false;
  static bool _isInitialized = false;
  static List<String>? _labels;
  static String? _initializationError;
  static bool _isUsingGPU = false;
  static int _optimalThreadCount = 2;

  // Model configuration from centralized config
  static const int modelInputSize = AIModelConfig.inputImageSize;
  static const int modelChannels = AIModelConfig.modelChannels;
  static const String modelPath = AIModelConfig.modelPath;
  static const String labelsPath = AIModelConfig.labelsPath;

  /// Initialize TensorFlow Lite model with optimized CPU/GPU detection
  static Future<bool> initialize() async {
    try {
      debugPrint('🔬 Initializing TensorFlow Lite model...');
      _isInitialized = true;
      _initializationError = null;

      // Detect optimal hardware configuration first
      await _detectOptimalConfiguration();

      // Check if model file exists first with detailed debugging
      try {
        debugPrint('🔍 Attempting to load model from: $modelPath');
        final modelData = await rootBundle.load(modelPath);
        debugPrint('📊 Model file found, size: ${modelData.lengthInBytes} bytes');

        // Verify the model data is valid
        if (modelData.lengthInBytes < 1000) {
          throw Exception('Model file too small, might be corrupted');
        }

        debugPrint('✅ Model file validation passed');
      } catch (e) {
        debugPrint('❌ Model file loading failed: $e');
        debugPrint('🔍 Error type: ${e.runtimeType}');
        debugPrint('⚠️ Model file not found or corrupted. Please ensure model.tflite is in assets/models/');
        debugPrint('📖 See assets/models/README.md for setup instructions');

        // Try to list available assets for debugging
        try {
          final manifestContent = await rootBundle.loadString('AssetManifest.json');
          debugPrint('📋 Available assets: $manifestContent');
        } catch (manifestError) {
          debugPrint('⚠️ Could not load asset manifest: $manifestError');
        }

        _initializationError = 'Model file not found. Please rebuild the app after ensuring the model file is in assets/models/';
        _isModelLoaded = false;
        await _loadLabels(); // Still load labels for UI
        return true; // Return true to allow app to continue
      }

      // Initialize interpreter with optimized configuration
      await _initializeInterpreter();

      // Load labels if available
      await _loadLabels();

      // If no labels were loaded from file, generate generic ones based on model output
      if (_labels == null) {
        _generateGenericLabels();
      }

      _isModelLoaded = true;
      debugPrint('✅ TensorFlow model loaded successfully');
      debugPrint('🚀 Using ${_isUsingGPU ? 'GPU' : 'CPU'} with $_optimalThreadCount threads');

      // Print model info
      _printModelInfo();

      return true;
    } catch (e) {
      debugPrint('❌ Failed to load TensorFlow model: $e');
      debugPrint('🔍 Error details: ${e.runtimeType}');
      _initializationError = 'Failed to initialize TensorFlow: $e';
      _isModelLoaded = false;
      await _loadLabels(); // Still load labels for UI
      return true; // Return true to allow app to continue
    }
  }

  /// Detect optimal hardware configuration for the device
  static Future<void> _detectOptimalConfiguration() async {
    try {
      // Detect CPU cores for optimal thread count
      _optimalThreadCount = Platform.numberOfProcessors;

      // Cap thread count based on device performance
      if (_optimalThreadCount > 4) {
        _optimalThreadCount = 4; // Most mobile GPUs work best with 4 threads max
      } else if (_optimalThreadCount < 2) {
        _optimalThreadCount = 2; // Minimum for decent performance
      }

      debugPrint('🔧 Detected $_optimalThreadCount CPU cores, using $_optimalThreadCount threads');

      // Check if we're running on a high-performance device
      final isHighPerformanceDevice = _optimalThreadCount >= 4;

      if (isHighPerformanceDevice && Platform.isAndroid) {
        debugPrint('📱 High-performance Android device detected, GPU delegate preferred');
      } else {
        debugPrint('📱 Standard device detected, CPU optimization preferred');
      }
    } catch (e) {
      debugPrint('⚠️ Hardware detection failed, using default configuration: $e');
      _optimalThreadCount = 2;
    }
  }

  /// Initialize interpreter with optimized GPU/CPU configuration
  static Future<void> _initializeInterpreter() async {
    // Strategy: Try GPU first on capable devices, then fallback to optimized CPU

    // First, try GPU delegate on Android devices
    if (Platform.isAndroid && _optimalThreadCount >= 3) {
      try {
        debugPrint('🚀 Attempting GPU initialization...');
        final gpuOptions = InterpreterOptions();

        // Configure GPU delegate with simplified settings
        final gpuDelegate = GpuDelegateV2(
          options: GpuDelegateOptionsV2(
            isPrecisionLossAllowed: false, // Keep precision for medical/agricultural use
          ),
        );

        gpuOptions.addDelegate(gpuDelegate);
        gpuOptions.threads = 1; // GPU delegate typically uses 1 thread

        _interpreter = await Interpreter.fromAsset(modelPath, options: gpuOptions);
        _isUsingGPU = true;

        debugPrint('✅ GPU delegate initialized successfully');
        return;
      } catch (e) {
        debugPrint('⚠️ GPU delegate initialization failed: $e');
        debugPrint('🔄 Falling back to optimized CPU execution...');
      }
    }

    // CPU-only execution with optimized settings
    try {
      debugPrint('🖥️ Initializing optimized CPU execution...');
      final cpuOptions = InterpreterOptions();

      // Optimize CPU execution
      cpuOptions.threads = _optimalThreadCount;

      _interpreter = await Interpreter.fromAsset(modelPath, options: cpuOptions);
      _isUsingGPU = false;

      debugPrint('✅ Optimized CPU execution initialized with $_optimalThreadCount threads');
    } catch (e) {
      // Final fallback with minimal settings
      debugPrint('⚠️ Optimized CPU failed, using basic CPU configuration: $e');
      final basicOptions = InterpreterOptions();
      basicOptions.threads = 2;

      _interpreter = await Interpreter.fromAsset(modelPath, options: basicOptions);
      _isUsingGPU = false;

      debugPrint('✅ Basic CPU execution initialized');
    }
  }

  /// Load model labels
  static Future<void> _loadLabels() async {
    try {
      final labelsData = await rootBundle.loadString(labelsPath);
      // Trim each label to remove \r, \n, and whitespace from Windows/Unix line endings
      _labels = labelsData
          .split('\n')
          .map((label) => label.trim())
          .where((label) => label.isNotEmpty)
          .toList();
      debugPrint('✅ Loaded ${_labels!.length} labels from file');
    } catch (e) {
      debugPrint('⚠️ Labels file not found, will generate generic labels after model loads: $e');
      _labels = null; // Will be generated after model loads
    }
  }

  /// Generate generic labels based on model output dimensions
  static void _generateGenericLabels() {
    if (_interpreter == null) {
      debugPrint('⚠️ Cannot generate labels: interpreter not available');
      return;
    }

    try {
      // Get the number of output classes from the model
      final outputTensor = _interpreter!.getOutputTensor(0);
      final outputShape = outputTensor.shape;

      // For classification models, the last dimension is usually the number of classes
      final numClasses = outputShape.last;

      // Generate generic labels: Class_0, Class_1, Class_2, etc.
      _labels = List.generate(numClasses, (index) => 'Class_$index');

      debugPrint('📋 Generated ${_labels!.length} generic labels based on model output shape: $outputShape');
      debugPrint('🏷️ Labels: ${_labels!.join(', ')}');
    } catch (e) {
      debugPrint('❌ Failed to generate generic labels: $e');
      // Fallback to minimal generic labels
      _labels = ['Class_0', 'Class_1'];
    }
  }

  /// Print model information
  static void _printModelInfo() {
    if (_interpreter == null) return;

    final inputTensors = _interpreter!.getInputTensors();
    final outputTensors = _interpreter!.getOutputTensors();

    debugPrint('📊 Model Info:');
    debugPrint('  Input tensors: ${inputTensors.length}');
    for (int i = 0; i < inputTensors.length; i++) {
      debugPrint(
          '    Input $i: ${inputTensors[i].shape} (${inputTensors[i].type})');
    }

    debugPrint('  Output tensors: ${outputTensors.length}');
    for (int i = 0; i < outputTensors.length; i++) {
      debugPrint(
          '    Output $i: ${outputTensors[i].shape} (${outputTensors[i].type})');
    }
  }

  /// Check if model is loaded
  static bool get isModelLoaded => _isModelLoaded && _interpreter != null;

  /// Check if service is initialized (even without model)
  static bool get isInitialized => _isInitialized;

  /// Get initialization error message
  static String? get initializationError => _initializationError;

  /// Debug method to check asset availability
  static Future<Map<String, dynamic>> debugAssetAvailability() async {
    debugPrint('🔍 Checking asset availability...');

    final results = <String, dynamic>{
      'modelFound': false,
      'modelSize': 0,
      'labelsFound': false,
      'labelsContent': '',
      'manifestFound': false,
      'modelInManifest': false,
      'allModelAssets': <String>[],
    };

    // Check model file
    try {
      final modelData =
          await rootBundle.load(modelPath);
      results['modelFound'] = true;
      results['modelSize'] = modelData.lengthInBytes;
      debugPrint('✅ Model file found: ${modelData.lengthInBytes} bytes');
    } catch (e) {
      debugPrint('❌ Model file not found: $e');
      debugPrint('🔍 Error details: ${e.toString()}');
    }

    // Check labels file
    try {
      final labelsData =
          await rootBundle.loadString(labelsPath);
      results['labelsFound'] = true;
      results['labelsContent'] = labelsData;
      debugPrint('✅ Labels file found: ${labelsData.length} characters');
      debugPrint('📋 Labels: ${labelsData.replaceAll('\n', ', ')}');
    } catch (e) {
      debugPrint('❌ Labels file not found: $e');
    }

    // Check asset manifest
    try {
      final manifestContent = await rootBundle.loadString('AssetManifest.json');
      final manifest = json.decode(manifestContent) as Map<String, dynamic>;
      results['manifestFound'] = true;

      final modelAssets = manifest.keys
          .where((key) => key.contains(modelPath.split('/').last))
          .toList();
      results['modelInManifest'] = modelAssets.isNotEmpty;
      debugPrint('🔍 Model assets in manifest: $modelAssets');

      final allModelAssets =
          manifest.keys.where((key) => key.contains('models/')).toList();
      results['allModelAssets'] = allModelAssets;
      debugPrint('🔍 All model assets: $allModelAssets');

      // Also check for any assets that might be similar
      final similarAssets = manifest.keys
          .where((key) =>
              key.toLowerCase().contains('apple') ||
              key.toLowerCase().contains('model'))
          .toList();
      debugPrint('🔍 Similar assets: $similarAssets');
    } catch (e) {
      debugPrint('❌ Could not check asset manifest: $e');
    }

    return results;
  }

  /// Check if service can be used (initialized, even without model)
  static bool get canAnalyze => _isInitialized;

  /// Analyze image for plant disease detection
  static Future<Map<String, dynamic>> analyzeImage(String imagePath) async {
    if (!_isInitialized) {
      return {
        'success': false,
        'error': 'TensorFlow service not initialized. Please restart the app.',
      };
    }

    if (!isModelLoaded || _interpreter == null) {
      // Provide demo results when model is not available
      debugPrint('🔬 Model not available, providing demo results');
      return _generateDemoResults(imagePath);
    }

    try {
      debugPrint('🔬 Starting TensorFlow analysis: $imagePath');

      // Preprocess image
      final inputData = await _preprocessImage(imagePath);
      if (inputData == null) {
        return {
          'success': false,
          'error': 'Failed to preprocess image',
        };
      }

      // Run inference
      final output = await _runInference(inputData);

      // Post-process results
      final results = _postprocessResults(output);

      debugPrint('✅ TensorFlow analysis completed');

      return {
        'success': true,
        'data': results,
        'analysisMethod': 'tflite_flutter',
      };
    } catch (e) {
      debugPrint('❌ TensorFlow analysis failed: $e');
      return {
        'success': false,
        'error': e.toString(),
      };
    }
  }

  /// Dispose resources and cleanup
  static Future<void> dispose() async {
    try {
      _interpreter?.close();
      _interpreter = null;
      _isModelLoaded = false;
      _isInitialized = false;
      _isUsingGPU = false;
      debugPrint('🧹 TensorFlow resources disposed');
    } catch (e) {
      debugPrint('⚠️ Error disposing TensorFlow resources: $e');
    }
  }

  /// Preprocess image for model input - OPTIMIZED VERSION
  /// Now runs in a background isolate to prevent UI freezing
  static Future<Float32List?> _preprocessImage(String imagePath) async {
    try {
      final stopwatch = Stopwatch()..start();

      // Run preprocessing in background isolate using compute()
      // This prevents UI freezing during heavy image processing
      final input = await compute(_preprocessImageInIsolate, {
        'imagePath': imagePath,
        'modelInputSize': modelInputSize,
        'modelChannels': modelChannels,
      });

      stopwatch.stop();

      if (input != null) {
        debugPrint('✅ Image preprocessed in ${stopwatch.elapsedMilliseconds}ms (background isolate): ${input.length} values');
      }

      return input;
    } catch (e) {
      debugPrint('❌ Image preprocessing failed: $e');
      return null;
    }
  }

  /// Optimized inference execution
  static Future<List<double>> _runInference(Float32List inputData) async {
    try {
      // Prepare input tensor with proper shape
      final input = inputData.reshape([1, modelInputSize, modelInputSize, modelChannels]);

      // Prepare output tensor
      final outputShape = _interpreter!.getOutputTensor(0).shape;
      final outputSize = outputShape.reduce((a, b) => a * b);
      final output = List.filled(outputSize, 0.0).reshape(outputShape);

      // Run inference with performance monitoring
      final stopwatch = Stopwatch()..start();
      _interpreter!.run(input, output);
      stopwatch.stop();

      // Log performance metrics
      final inferenceTime = stopwatch.elapsedMilliseconds;
      debugPrint('⚡ Inference completed in ${inferenceTime}ms using ${_isUsingGPU ? 'GPU' : 'CPU'}');

      // Extract and validate results
      final results = output[0] as List<double>;

      // Validate output
      if (results.any((value) => value.isNaN || value.isInfinite)) {
        debugPrint('⚠️ Invalid inference results detected');
        throw Exception('Invalid inference output');
      }

      debugPrint('🔬 Inference results: ${results.take(5).toList()}... (showing first 5)');
      return results;
    } catch (e) {
      debugPrint('❌ Inference execution failed: $e');
      rethrow;
    }
  }

  /// Post-process model results
  static Map<String, dynamic> _postprocessResults(List<double> rawResults) {
    if (_labels == null || _labels!.isEmpty) {
      return {
        'predictions': [],
        'topPrediction': null,
        'confidence': 0.0,
      };
    }

    // Create predictions with labels and confidence scores
    final predictions = <Map<String, dynamic>>[];
    for (int i = 0; i < rawResults.length && i < _labels!.length; i++) {
      final confidence = rawResults[i];
      final label = _labels![i];

      predictions.add({
        'label': label,
        'confidence': confidence,
        'displayName': _formatLabel(label),
      });
    }

    // Sort by confidence (descending)
    predictions.sort((a, b) =>
        (b['confidence'] as double).compareTo(a['confidence'] as double));

    // Get top prediction
    final topPrediction = predictions.isNotEmpty ? predictions.first : null;

    // Build detectedDiseases list for storage
    // Only store 'label' and 'confidence' - no need for redundant 'disease' field
    List<Map<String, dynamic>> detectedDiseases = [];
    if (predictions.isNotEmpty) {
      for (final pred in predictions) {
        detectedDiseases.add({
          'label': pred['label'],
          'confidence': pred['confidence'],
        });
      }
    } else {
      detectedDiseases.add({
        'label': 'Unknown',
        'confidence': 0.0,
      });
    }

    return {
      'predictions': predictions,
      'topPrediction': topPrediction,
      'confidence': topPrediction?['confidence'] ?? 0.0,
      'isHealthy': topPrediction?['label']?.contains('healthy') ?? false,
      'diseaseDetected': topPrediction != null &&
          !(topPrediction['label']?.contains('healthy') ?? true),
      'detectedDiseases': detectedDiseases,
    };
  }

  /// Format label for display
  static String _formatLabel(String label) {
    // Convert "Apple___Apple_scab" to "Apple Scab"
    return label
        .split('___')
        .last
        .replaceAll('_', ' ')
        .split(' ')
        .map((word) => word.isNotEmpty
            ? '${word[0].toUpperCase()}${word.substring(1).toLowerCase()}'
            : '')
        .join(' ');
  }

  /// Get model labels
  static List<String> get labels => _labels ?? [];

  /// Generate demo results when model is not available
  static Map<String, dynamic> _generateDemoResults(String imagePath) {
    // Use actual labels if available, otherwise generate generic ones
    List<String> availableLabels;

    if (_labels != null && _labels!.isNotEmpty) {
      // Use the actual labels from the model
      availableLabels = _labels!;
    } else {
      // Generate generic labels for demo (simulating a typical classification model)
      availableLabels = List.generate(5, (index) => 'Class_$index');
    }

    final predictions = <Map<String, dynamic>>[];

    // Create predictions for all available labels with pseudo-random but realistic confidences
    for (int i = 0; i < availableLabels.length; i++) {
      final label = availableLabels[i];
      
      // Generate deterministic pseudo-random confidence based on image path and label index
      final seed = (imagePath.hashCode + label.hashCode + i) % 100;
      final baseConfidence = (seed / 100.0) * 0.8 + 0.1; // Range: 0.1 to 0.9
      
      // Apply some variation to make it more realistic
      final variation = ((seed * 3) % 20 - 10) / 100.0; // ±10% variation
      final confidence = (baseConfidence + variation).clamp(0.05, 0.95);

      predictions.add({
        'label': label,
        'confidence': confidence,
        'displayName': _formatLabel(label),
      });
    }

    // Sort by confidence (descending) to simulate real model behavior
    predictions.sort((a, b) =>
        (b['confidence'] as double).compareTo(a['confidence'] as double));

    final topPrediction = predictions.first;

    // Build detectedDiseases list
    final detectedDiseases = predictions.map((pred) => {
      'disease': pred['displayName'],
      'confidence': pred['confidence'],
      'label': pred['label'],
    }).toList();

    // Determine if the top prediction suggests a healthy state
    // This is generic - looks for keywords that might indicate health
    final topLabel = (topPrediction['label'] as String).toLowerCase();
    final isHealthy = topLabel.contains('healthy') ||
                     topLabel.contains('normal') ||
                     topLabel.contains('class_0'); // Often class 0 is healthy in many models

    return {
      'success': true,
      'data': {
        'predictions': predictions,
        'topPrediction': topPrediction,
        'confidence': topPrediction['confidence'],
        'isHealthy': isHealthy,
        'diseaseDetected': !isHealthy,
        'detectedDiseases': detectedDiseases,
        'isDemoResult': true, // Important: mark this as demo data
      },
      'analysisMethod': 'demo_fallback',
    };
  }
}

File: src/services/TensorFlowService.js
import { loadTensorflowModel } from 'react-native-fast-tflite';

/**
 * TensorFlow Lite Service for Plant Disease Detection
 * CORRECTED IMPLEMENTATION - Based on Python app_basic.py
 */
class TensorFlowService {
	static model = null;
	static isModelLoaded = false;
	static modelPath = require('../../assets/apple_model_final.tflite');

	// CORRECTED: Match Python model configuration exactly
	static MODEL_INPUT_SIZE = 128; // Changed from 224 to 128 to match Python
	static MODEL_CHANNELS = 3;

	/**
	 * Load the TensorFlow Lite model
	 * @returns {Promise<boolean>} Success status
	 */
	static async loadModel() {
		try {
			if (this.isModelLoaded && this.model) {
				console.log('✅ Model already loaded');
				return true;
			}

			console.log('🔄 Loading TensorFlow Lite model...');
			this.model = await loadTensorflowModel(this.modelPath);

			if (!this.model) {
				throw new Error('Model loading returned null or undefined');
			}

			this.isModelLoaded = true;
			console.log('✅ TensorFlow Lite model loaded successfully');
			console.log(
				'📊 Model input size:',
				this.MODEL_INPUT_SIZE + 'x' + this.MODEL_INPUT_SIZE,
			);
			return true;
		} catch (error) {
			console.error('❌ Error loading TensorFlow Lite model:', error);
			this.isModelLoaded = false;
			this.model = null;
			return false;
		}
	}

	/**
	 * Process image using react-native-fast-tflite (REAL IMPLEMENTATION)
	 * This completely removes mock data and uses the official library
	 * @param {string} imageUri - Image URI
	 * @returns {Promise<Object>} Analysis results
	 */
	static async processImageWithFastTflite(imageUri) {
		try {
			console.log(
				'🔄 Processing image with react-native-fast-tflite:',
				imageUri,
			);

			if (!this.isModelLoaded || !this.model) {
				const loaded = await this.loadModel();
				if (!loaded) {
					throw new Error('Failed to load TensorFlow Lite model');
				}
			}

			// Use react-native-fast-tflite's built-in image processing
			// This handles all the preprocessing automatically
			console.log('🚀 Running inference with react-native-fast-tflite');
			const results = await this.model.run([imageUri]);

			console.log('✅ Inference completed successfully');
			console.log('📊 Raw results:', results);

			return results;
		} catch (error) {
			console.error('❌ react-native-fast-tflite processing failed:', error);
			throw error;
		}
	}

	/**
	 * Process results exactly like Python code
	 * @param {Object|Array} rawResults - Raw model output
	 * @returns {Object} Processed results
	 */
	static postprocessResults(rawResults) {
		try {
			console.log('🔍 Processing results (Python-style)...');
			console.log('📊 Raw results:', rawResults);

			// Extract output tensor (matching Python: output_data = interpreter.get_tensor(...))
			let outputData = null;

			if (Array.isArray(rawResults)) {
				outputData = rawResults[0];
			} else if (rawResults && typeof rawResults === 'object') {
				const keys = Object.keys(rawResults);
				if (keys.length > 0) {
					outputData = rawResults[keys[0]];
				}
			}

			if (!outputData) {
				throw new Error('Could not extract output data from results');
			}

			// Convert to array if needed
			if (!Array.isArray(outputData)) {
				outputData = Array.from(outputData);
			}

			console.log('📊 Extracted output data:', outputData);

			// EXACT Python logic: predicted_class_index = np.argmax(output_data)
			let maxValue = -Infinity;
			let predictedClassIndex = 0;

			for (let i = 0; i < outputData.length; i++) {
				if (outputData[i] > maxValue) {
					maxValue = outputData[i];
					predictedClassIndex = i;
				}
			}

			// EXACT Python logic: confidence = float(np.max(output_data))
			const confidence = maxValue;

			// EXACT Python class names (matching Python: CLASS_NAMES)
			const classNames = [
				'Apple___Apple_scab',
				'Apple___Black_rot',
				'Apple___Cedar_apple_rust',
				'Apple___healthy',
			];

			const displayNames = [
				'Apple Scab',
				'Apple Black Rot',
				'Cedar Apple Rust',
				'Healthy',
			];

			// EXACT Python logic: predicted_class_name = CLASS_NAMES[predicted_class_index]
			const predictedClassName = classNames[predictedClassIndex];
			const predictedDisplayName = displayNames[predictedClassIndex];

			console.log('🎯 PYTHON-STYLE PREDICTION:');
			console.log('   - Class Index:', predictedClassIndex);
			console.log('   - Class Name:', predictedClassName);
			console.log('   - Confidence:', confidence);
			console.log('   - All outputs:', outputData);

			// Create detected diseases
			const detectedDiseases = [];

			if (predictedClassName !== 'Apple___healthy') {
				detectedDiseases.push({
					id: predictedClassName,
					name: predictedDisplayName,
					className: predictedClassName,
					confidence: confidence,
					severity:
						confidence > 0.7 ? 'High' : confidence > 0.4 ? 'Medium' : 'Low',
					description: `Detected ${predictedDisplayName} with confidence ${(
						confidence * 100
					).toFixed(1)}%`,
				});
			}

			const isHealthy = predictedClassName === 'Apple___healthy';

			return {
				plantType: 'Apple',
				cropId: 1,
				diseases: detectedDiseases,
				healthStatus: isHealthy ? 'Healthy' : 'Diseased',
				overallConfidence: confidence,
				prediction: {
					className: predictedClassName,
					displayName: predictedDisplayName,
					confidence: confidence,
					index: predictedClassIndex,
				},
				allProbabilities: outputData.map((prob, idx) => ({
					className: classNames[idx] || `Class_${idx}`,
					displayName: displayNames[idx] || `Class ${idx}`,
					probability: prob,
					percentage: (prob * 100).toFixed(1) + '%',
				})),
				rawPredictions: Array.from(outputData),
				modelUsed: 'TensorFlow Lite (Python-Compatible)',
				timestamp: new Date().toISOString(),
			};
		} catch (error) {
			console.error('❌ Error in postprocessing:', error);
			throw error;
		}
	}

	/**
	 * Analyze image using Python-compatible pipeline
	 * @param {string} imageUri - URI of the image to analyze
	 * @returns {Promise<Object>} Analysis results
	 */
	static async analyzeImage(imageUri) {
		try {
			console.log(
				'🔬 Starting REAL TensorFlow analysis with react-native-fast-tflite:',
				imageUri,
			);

			// Use the official react-native-fast-tflite library
			// This handles all image preprocessing automatically
			const rawResults = await this.processImageWithFastTflite(imageUri);

			// Process results using the same logic as Python
			const results = this.postprocessResults(rawResults);

			console.log('✅ Real TensorFlow analysis completed:', results);

			return {
				success: true,
				data: results,
				analysisMethod: 'react_native_fast_tflite_real',
			};
		} catch (error) {
			console.error('❌ Real TensorFlow analysis failed:', error);
			return {
				success: false,
				error: error.message || 'TensorFlow analysis failed',
				analysisMethod: 'react_native_fast_tflite_error',
			};
		}
	}

	/**
	 * Get model information
	 */
	static getModelInfo() {
		return {
			modelPath: this.modelPath,
			isLoaded: this.isModelLoaded,
			modelType: 'TensorFlow Lite (Python-Compatible)',
			framework: 'react-native-fast-tflite',
			inputSize: this.MODEL_INPUT_SIZE, // 128
			channels: this.MODEL_CHANNELS, // 3
			supportedClasses: [
				'Apple___Apple_scab',
				'Apple___Black_rot',
				'Apple___Cedar_apple_rust',
				'Apple___healthy',
			],
		};
	}

	/**
	 * Reset the model
	 */
	static resetModel() {
		this.model = null;
		this.isModelLoaded = false;
		console.log('Model reset');
	}

	/**
	 * Check if model is ready
	 */
	static isReady() {
		return this.isModelLoaded && this.model !== null;
	}
}

export default TensorFlowService;
